TYP: incorrect type inference for iteration elements? (numpy 2.3.0)

### Describe the issue:

When updating from numpy 2.2.6 to 2.3.0, I encountered a new type check error that I believe is incorrect.

When you iterate over a matrix, the elements should be the same shape as the original matrix, but with one dimension dropped. Because the `NDArray` type has ambiguous shape, elements could either be **an array OR a scalar**.

However, in numpy 2.3.0, when I iterate over a matrix of type `NDArray[np.float64]`, numpy typing says each element has type `np.float64`. It is assuming that the matrix is 1-dimensional and each element is a scalar, which is incorrect.

```py
arr: NDArray[np.float64] = np.zeros((5,3))
for element in arr:
  # Each `element` is an `NDArray[np.float64]` of shape (3)
  # But instead, numpy typing says it is a `float64`
```

In numpy 2.2.6, each element had type `Any` - vague, but at least not incorrect.

---

From the release notes, I see that improvements were made to numpy's static typing in 2.3.0. I also see `mypy_plugin` was deprecated, but I don't think that's the issue because I didn't have the plugin enabled in the first place.

So is this an actual bug, or am I misunderstanding how to use static typing with numpy? I feel a little shaky on the subject. If I am doing this wrong, please show me the correct way 🙏 

### Reproduce the code example:

```python
"""A simple example that should throw a type check error."""

import numpy as np
from numpy.typing import NDArray


def example(pixels: NDArray[np.float64]) -> None:
	"""`pixels` is an NDArray of shape (length, 3)."""
	# Convert NDArray to a list of tuples
	tuples: list[tuple] = [tuple(pixel) for pixel in pixels]
```

### Error message:

```shell
error: Argument 1 to "tuple" has incompatible type "float64"; expected "Iterable[Any]"  [arg-type]
```

### Python and NumPy Versions:

```
2.3.0
3.13.5 (main, Jun 12 2025, 22:47:08) [GCC 12.2.0]
```

### Type-checker version and settings:

mypy version: 1.16.1
mypy command: `mypy .`

### Additional typing packages.

_No response_

--- a/numpy/__init__.pyi
+++ b/numpy/__init__.pyi
@@ -807,6 +807,7 @@ _RealNumberT = TypeVar("_RealNumberT", bound=floating | integer)
 _FloatingT_co = TypeVar("_FloatingT_co", bound=floating, default=floating, covariant=True)
 _IntegerT = TypeVar("_IntegerT", bound=integer)
 _IntegerT_co = TypeVar("_IntegerT_co", bound=integer, default=integer, covariant=True)
+_NonObjectScalarT = TypeVar("_NonObjectScalarT", bound=np.bool | number | flexible | datetime64 | timedelta64)
 
 _NBit = TypeVar("_NBit", bound=NBitBase, default=Any)  # pyright: ignore[reportDeprecated]
 _NBit1 = TypeVar("_NBit1", bound=NBitBase, default=Any)  # pyright: ignore[reportDeprecated]
@@ -2572,10 +2573,13 @@ class ndarray(_ArrayOrScalarCommon, Generic[_ShapeT_co, _DTypeT_co]):
     def __len__(self) -> int: ...
     def __contains__(self, value: object, /) -> builtins.bool: ...
 
-    @overload  # == 1-d & object_
-    def __iter__(self: ndarray[tuple[int], dtype[object_]], /) -> Iterator[Any]: ...
-    @overload  # == 1-d
-    def __iter__(self: ndarray[tuple[int], dtype[_ScalarT]], /) -> Iterator[_ScalarT]: ...
+    # NOTE: This weird `Never` tuple works around a strange mypy issue where it assigns
+    # `tuple[int]` to `tuple[Never]` or `tuple[int, int]` to `tuple[Never, Never]`.
+    # This way the bug only occurs for 9-D arrays, which are probably not very common.
+    @overload
+    def __iter__(self: ndarray[tuple[Never, Never, Never, Never, Never, Never, Never, Never, Never]], /) -> Iterator[Any]: ...
+    @overload  # == 1-d & dtype[T \ object_]
+    def __iter__(self: ndarray[tuple[int], dtype[_NonObjectScalarT]], /) -> Iterator[_NonObjectScalarT]: ...
     @overload  # >= 2-d
     def __iter__(self: ndarray[tuple[int, int, *tuple[int, ...]], dtype[_ScalarT]], /) -> Iterator[NDArray[_ScalarT]]: ...
     @overload  # ?-d

--- a/numpy/typing/tests/data/reveal/ndarray_misc.pyi
+++ b/numpy/typing/tests/data/reveal/ndarray_misc.pyi
@@ -6,6 +6,7 @@ function-based counterpart in `../from_numeric.py`.
 
 """
 
+from collections.abc import Iterator
 import ctypes as ct
 import operator
 from types import ModuleType
@@ -28,6 +29,10 @@ AR_c8: npt.NDArray[np.complex64]
 AR_m: npt.NDArray[np.timedelta64]
 AR_U: npt.NDArray[np.str_]
 AR_V: npt.NDArray[np.void]
+
+AR_f8_1d: np.ndarray[tuple[int], np.dtype[np.float64]]
+AR_f8_2d: np.ndarray[tuple[int, int], np.dtype[np.float64]]
+AR_f8_3d: np.ndarray[tuple[int, int, int], np.dtype[np.float64]]
 
 ctypes_obj = AR_f8.ctypes
 
@@ -235,3 +240,8 @@ assert_type(AR_m.to_device("cpu"), npt.NDArray[np.timedelta64])
 
 assert_type(f8.__array_namespace__(), ModuleType)
 assert_type(AR_f8.__array_namespace__(), ModuleType)
+
+assert_type(iter(AR_f8), Iterator[Any])  # any-D
+assert_type(iter(AR_f8_1d), Iterator[np.float64])  # 1-D
+assert_type(iter(AR_f8_2d), Iterator[npt.NDArray[np.float64]])  # 2-D
+assert_type(iter(AR_f8_3d), Iterator[npt.NDArray[np.float64]])  # 3-D

